### PR TITLE
New best weights after supervised learning from large pgn files using optimize with generator

### PR DESCRIPTION
**Pull request by brianprichardson**
Trained for about four days using supervised learning feeding large pgn files (CCRL) directly to new version of optimize.py with fit_generator.  

Did several passes reducing learning rate from .01 to .001 and finally .0001  
Used batch size of 2,048 and increased number of steps to 100,000 for LR .0001 (that took 3 days).  

Win rate v prior best weights 96.6% after 29 games.

### CODE REVIEW

**Comment by Zeta36** [comment]
Great!! Thank you very much for your effort, @benediamond. Do you think there is still place for better results with our model?

**Comment by Akababa** [comment]
Nice job! I find that the opening overfit is less pronounced but surprisingly it still gives away material in the middlegame and endgame. Have you had a chance to compare it to leela-chess?

**Comment by brianprichardson** [comment]
First big thanks to @Zeta36 , @Akababa , @benediamond, (and any others I am overlooking), for all of your work and sharing it.  Yes, I am waiting for leela-chess to really get going (nearly there).  In parallel with that, I am trying a somewhat modified approach.  My thinking is that we simply don't have the massive hardware resources that alpha-zero-chess used.  So, instead of starting tabula rasa, I am trying to use millions of games that have already been played.  Using some material adjudication and EGTBs are also options.  Ultimately, I would like flexibility in being able to easily change the NN input (like using some or all of the 8 prior moves, with or without threefold repetitions, castling options, etc), along with of course, changing the NN model (how deep, etc).

Since on my PCs it takes so long to generate the SL play data files and they take a large amount of space, I thought just using the pgn files as input might be a good idea.  So, I am using data generators (and currently adding validation).  With a quite consistent 140 average move plies per game, the 100,000 steps of 2,048 in the new best weights above represent about 1.4 million games.  This took 3 days (after the much faster 1,000 and 10,000 step runs with LR .01 and .001).  Before ramping up to many more games, there are some constraints that need to be addressed.

I noticed that my GTX1070 was not very busy, so a lot of the time seems to be spent in preparing the sample batches.  I am still relatively new to Python (so concatenate and slicing takes me hours to figure out), and am looking at how to speed things up with multiple fit_generator queues and workers (incidentally, I have taken out the threads for SL and self play)  In addition, I am adding graceful pausing and stopping (using https://github.com/samyzaf/kerutils with modifications for multiple outputs and Python3).  I also have a crude "restart" capability just using game offsets (per python-chess).

All of this will hopefully give me some monitoring tools to see when a model is over/under fitting, when to add more games (input samples), etc.  Right now another run with 20 3x3 is up to 50,000 steps.  Then, I have to figure out how to run evaluation matches (using cutechess or xboard) to have different models (not just the same model with different weights) play each other.  So, the first step will be comparing the 7 layer model with the 20 layer one.  I think I can shorten the 3-days by quite a lot too (at least 4x).

Finally, back to LCZ, I am quite familiar with Stockfish, but less so with TF in C++, so I prefer things at the Keras framework level.  Another future speed boost could come from using Stockfish C++ pgn routines with Python Keras training.  As the author of a mediocre chess engine (Tinker), the chess engine side of things is not the issue, it just takes me a while to figure out the Python (the Python 2 to 3 transition was/is painful, and still learning the basics like lists v tuples and numpy, iterators v generators, etc) and ML (moving targets there too as I started with FANN to learn just piece values, then Keras/Theano on Win7, then went to Keras/TF when Windows support got better) side of things, and broken backwards compatibility has cost me a lot of time.  I have gradually transitioned from only Windows to also using Ubuntu (but still just a Linux noob).    I can usually tweak something to incrementally add to it.  Like now I want to learn how to add to the config (mini, normal) files and pass arguments that way instead of hard coding them in.

Any comments and suggestions would be welcome.

Thanks,
Brian

**Comment by benediamond** [comment]
@brianprichardson, @Zeta36, @Akababa Awesome work, everyone! i've now gotten a job---which is why my account has gone dead---but I'm glad you all are continuing to advance this work. Hopefully i'll be able to get back to it all some day. Best of luck.

**Comment by brianprichardson** [comment]
> @Akababa
 Nice job! I find that the opening overfit is less pronounced but surprisingly it still gives away material in the middlegame and endgame. Have you had a chance to compare it to leela-chess?
> 

I have a notion about why it gives away material.  So far, the games are all computer v computer, and engines never drop material within their search horizons unless forced to.  Thus, there are almost no samples from which to learn.  I am thinking of doing the training in phases starting with lower-ranked players that make blunders, and then increasing the rating range.
--

**Comment by brianprichardson** [comment]
@Akababa
Nice job! I find that the opening overfit is less pronounced but surprisingly it still gives away material in the middlegame and endgame. Have you had a chance to compare it to leela-chess?

I have a notion about why it gives away material. So far, the games are all computer v computer, and engines never drop material within their search horizons unless forced to. Thus, there are almost no samples from which to learn. I am thinking of doing the training in phases starting with lower-ranked players that make blunders, and then increasing the rating range.

**Comment by brianprichardson** [comment]
@matteopincherle Well, I don't recall exactly, as I am still not being systematic and taking notes for fully reproducible results.  In case you missed it in the earlier post, one of the key things is that I modified optimize.py to feed the training using a generator instead of the processed output of the "sl".  I used a CCRL 40/4 file with about 1.5M games, but I don't think it got thru them all.  The runs with learning rates of  .01 and .001 took about a day, and then there was a long run for 3 days at .0001  The PC is an old i920 overclocked to 4GHz with 24GB RAM and an 8GB GTX1070 gpu, so while old, it is still moderately fast, IMO.  Since this weights update, I have modified it further to use int8 input to the NN, which cuts RAM use a ton.  I am trying to use raw pgn input as I want to be able to experiment with things like the number of history planes, and it seems like "big data" FTW ("for the win") with NNs, at least from what I've read.  Of course, Leela Chess Zero is taking a more "pure" tabula rasa approach, but with minuscule h/w relative to Google, I am going the supervised route.  I am also currently looking at Nadam and ELU.  I'm still pretty much a noob with Ubuntu, Python, and NNs, so I'm learning as I go.  It took a whole day to figure out how to superimpose my blue testing graph on the orange training runs in tensorboard.  Another day to figure out the lambda for converting the int8 input to float32 on the fly, you get the idea.
